lib/_sendEvent.ts

```diff
@@ -16,98 +17,117 @@ export type InsightsEvent = {
-  }
-
-  const event: InsightsEvent = {
-    eventType,
-    eventName: eventData.eventName,
-    userToken: eventData.userToken || this._userToken,
-    index: eventData.index
-  };
+export function makeSendEvent(requestFn: RequestFnType) {
+  return function sendEvent(
+    eventType: InsightsEventType,
+    eventData: InsightsEvent
+  ) {
+    if (this._userHasOptedOut) {
+      return;
+    }
+    if (!this._hasCredentials) {
+      throw new Error(
+        "Before calling any methods on the analytics, you first need to call the 'init' function with appId and apiKey parameters"
+      );
+    }
+    const userToken = eventData.userToken || this._userToken;
+    if (userToken === undefined) {
+      throw new Error(
+        "Before calling any methods on the analytics, you first need to call 'setUserToken' function or include 'userToken' in the event payload."
+      );
+    }
+
+    // mandatory params
+    if (!isString(eventData.index)) {
+      throw TypeError("expected required parameter `index` to be a string");
+    }
+    if (!isString(eventData.eventName)) {
+      throw TypeError("expected required parameter `eventName` to be a string");
+    }
 
-  // optional params
-  if (!isUndefined(eventData.timestamp)) {
-    if (!isNumber(eventData.timestamp)) {
-      throw TypeError("expected optional parameter `timestamp` to be a number");
+    if (!isUndefined(eventData.userToken) && !isString(eventData.userToken)) {
+      throw TypeError("expected optional parameter `userToken` to be a string");
     }
-    event.timestamp = eventData.timestamp;
-  }
 
-  if (!isUndefined(eventData.queryID)) {
-    if (!isString(eventData.queryID)) {
-      throw TypeError("expected optional parameter `queryID` to be a string");
+    const event: InsightsEvent = {
+      eventType,
+      eventName: eventData.eventName,
+      userToken,
+      index: eventData.index
+    };
+
+    // optional params
+    if (!isUndefined(eventData.timestamp)) {
+      if (!isNumber(eventData.timestamp)) {
+        throw TypeError(
+          "expected optional parameter `timestamp` to be a number"
+        );
+      }
+      event.timestamp = eventData.timestamp;
     }
-    event.queryID = eventData.queryID;
-  }
 
-  if (!isUndefined(eventData.objectIDs)) {
-    if (!Array.isArray(eventData.objectIDs)) {
-      throw TypeError("expected optional parameter `objectIDs` to be an array");
+    if (!isUndefined(eventData.queryID)) {
+      if (!isString(eventData.queryID)) {
+        throw TypeError("expected optional parameter `queryID` to be a string");
+      }
+      event.queryID = eventData.queryID;
     }
-    event.objectIDs = eventData.objectIDs;
-  }
 
-  if (!isUndefined(eventData.positions)) {
-    if (!Array.isArray(eventData.positions)) {
-      throw TypeError("expected optional parameter `positions` to be an array");
+    if (!isUndefined(eventData.objectIDs)) {
+      if (!Array.isArray(eventData.objectIDs)) {
+        throw TypeError(
```

Comment:
```suggestion
        throw new TypeError(
```

Reply:
do a global search in the project to fix this for `throw `, and add new everywhere